src/dftbp/common/memman.F90

```diff
@@ -93,43 +90,76 @@ subroutine TAlignedArray_allocateAligned(this, size, alignment)
     !> Alignment (default: 64 bytes)
     integer, intent(in), optional :: alignment
 
-    if (this%isAllocated) then
+    integer :: dp_size
+
+    if (c_associated(this%memoryPointer)) then
       call error("Aligned array is already allocated")
     end if
 
     @:ASSERT(size > 0)
 
+    dp_size = int(c_sizeof(0._dp))
     if (present(alignment)) then
-      @:ASSERT(alignment >= dp)
-      @:ASSERT(mod(alignment, dp) == 0)
+      @:ASSERT(alignment >= dp_size)
+      @:ASSERT(mod(alignment, dp_size) == 0)
 
       this%alignment = alignment
     end if
 
     if (posixMemalign(this%memoryPointer, int(this%alignment, kind=c_size_t),&
-          & int(size * dp, kind=c_size_t)) /= 0) then
+          & int(size * dp_size, kind=c_size_t)) /= 0) then
       call error("Error during allocation of aligned array")
     end if
 
+    @:ASSERT(c_associated(this%memoryPointer))
+
     this%size = size
-    call c_f_pointer(this%memoryPointer, this%array, [size])
 
-    this%isAllocated = .true.
+  end subroutine TAlignedArray_allocate
+
+
+  !> Fetch the pointer to the array data
+  subroutine TAlignedArray_getArray(this, array)
```

Comment:
Yes, this is definitely the most robust solution. One stylistic remark (open for discussion): one could also implement this as function, so that on the caller side you had 
```
myArray => aligned%array()
```
instead of 
```
call aligned%getArray(myArray)
```.
I like the first one more from the optics, but it is more error prone, as if somebody writes `=` instead of `=>`, it will warrantedly crash at run-time. So, probably let's stay with the subroutine....

Reply:
I was also thinking about that, because we are using it in ELPA. However, I do not like it, because of the exact reason you have mentioned: If the user just writes a `=`, then everything gets messed up. The subroutine is safer for sure.